usim/_basics/resource.py

```diff
@@ -0,0 +1,208 @@
+from typing import TypeVar, Generic, Optional, Type
+
+from .._core.loop import __LOOP_STATE__
+from ._resource_level import __specialise__, ResourceLevels
+from .tracked import Tracked
+
+
+T = TypeVar('T')
+
+
+class BaseResources(Generic[T]):
+    """
+    Internal base class for resource types
+    """
+    _levels_type = None  # type: Type[ResourceLevels[T]]
+    _available = None  # type: Tracked[ResourceLevels[T]]
+
+    @property
+    def levels(self) -> ResourceLevels[T]:
+        """Current levels of resources"""
+        return self._available.value
+
+    @property
+    def resources_type(self) -> Type[ResourceLevels[T]]:
+        """Type of underlying resources"""
+        return self._levels_type
+
+    async def __insert_resources__(self, amounts: ResourceLevels):
+        new_levels = self._available.value + amounts
+        await self._available.set(new_levels)
+
+    async def __remove_resources__(self, amounts: ResourceLevels):
+        new_levels = self._available.value - amounts
+        await self._available.set(new_levels)
+
+    def borrow(self, **amounts: T) -> 'BorrowedResources[T]':
+        """
+        Temporarily borrow resources for a given context
+
+        :param amounts: resource levels to borrow
+        :return: async context to borrow resources
+        """
+        borrowed_levels = self._levels_type(**amounts)
+        assert self._levels_type.zero <= borrowed_levels,\
+            'cannot borrow negative amounts'
+        return BorrowedResources(self, borrowed_levels)
+
+
+class BorrowedResources(BaseResources[T]):
+    """
+    Fixed supply of named resources temporarily taken from another resource supply
+    """
+    @property
+    def _levels_type(self):
+        return self._resources._levels_type
+
+    def __init__(self, resources: 'BaseResources', capacity: ResourceLevels):
+        self._resources = resources
+        self._capacity = capacity
+        self._available = Tracked(self._levels_type.zero)
+
+    async def __aenter__(self):
+        await (self._resources._available >= self._capacity)
+        await self._resources.__remove_resources__(self._capacity)
+        await self.__insert_resources__(self._capacity)
+        return self
+
+    async def __aexit__(self, exc_type, exc_val, exc_tb):
+        if exc_type is GeneratorExit:
+            # we are killed forcefully and cannot perform async operations
+            # dispatch a new activity to release our resources eventually
+            __LOOP_STATE__.LOOP.schedule(
+                self._resources.__insert_resources__(self._capacity)
+            )
+            __LOOP_STATE__.LOOP.schedule(
+                self.__remove_resources__(self._capacity)
+            )
```

Comment:
We should consider changing the order of operations. If something happens in removing `capacity` from `available` resource, we have doubled the resources. So I would rather first remove the resources from `available` before adding them back again to `resources`.